server/src/test/java/org/opensearch/extensions/ExtensionsManagerTests.java

```diff
@@ -298,12 +309,74 @@ public void testNonUniqueExtensionsDiscovery() throws Exception {
                     "fakeClass1",
                     new ArrayList<String>(),
                     false
-                )
+                ),
+                Collections.emptyList()
             )
         );
         assertEquals(expectedExtensionsList.size(), extensionsManager.getExtensionIdMap().values().size());
         assertTrue(expectedExtensionsList.containsAll(extensionsManager.getExtensionIdMap().values()));
         assertTrue(extensionsManager.getExtensionIdMap().values().containsAll(expectedExtensionsList));
+        assertTrue(expectedExtensionsList.containsAll(emptyList()));
+    }
+
+    public void testDiscoveryExtension() throws Exception {
+        String expectedId = "test id";
+        Version expectedVersion = Version.fromString("2.0.0");
+        ExtensionDependency expectedDependency = new ExtensionDependency(expectedId, expectedVersion);
+
+        DiscoveryExtensionNode discoveryExtension = new DiscoveryExtensionNode(
+            "firstExtension",
+            "uniqueid1",
+            "uniqueid1",
+            "myIndependentPluginHost1",
+            "127.0.0.0",
+            new TransportAddress(InetAddress.getByName("127.0.0.0"), 9300),
+            new HashMap<String, String>(),
+            Version.fromString("3.0.0"),
+            new PluginInfo(
+                "firstExtension",
+                "Fake description 1",
+                "0.0.7",
+                Version.fromString("3.0.0"),
+                "14",
+                "fakeClass1",
+                new ArrayList<String>(),
+                false
+            ),
+            List.of(expectedDependency)
+        );
+
+        assertEquals(List.of(expectedDependency), discoveryExtension.getDependencies());
+
+        try (BytesStreamOutput out = new BytesStreamOutput()) {
+            discoveryExtension.writeTo(out);
+            out.flush();
+            try (BytesStreamInput in = new BytesStreamInput(BytesReference.toBytes(out.bytes()))) {
+                discoveryExtension = new DiscoveryExtensionNode(in);
+
+                assertEquals(List.of(expectedDependency), discoveryExtension.getDependencies());
```

Comment:
Shouldn't we compare dependency by elements here rather than list directly?

Reply:
The implementation of `ArrayList.equals()` does exactly that.

Although there is an `assertIterableEquals()` method in Junit that would work here as well.

Reply:
List option as-is:  checks all items in the list and probably better from a "test it all" stand point, but relies on implementation details.

List option with `assertIterableEquals()`: identical in practice but doesn't rely on assumptions of the implementation of `List.equals()`.

Extracting a single item:   relies on what we know about the list and the single element, and if we later decide we want to test two elements, would have to be rewritten.

I prefer the second option but am ok with the first one. I don't like the third but really all are equivalent with this test.

Reply:
Seems [`AssertIterableEquals`](https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/Assertions.html#assertIterableEquals-java.lang.Iterable-java.lang.Iterable-) is a JUnit 5 feature.  Not sure why, but this repo is still on JUnit 4 so that won't work.

Personally I think it's fine as is, given the [`ArrayList` `equals()`](https://github.com/openjdk/jdk/blob/8ea2a6777b68986df9d191f1bf983549d72cb3f8/src/java.base/share/classes/java/util/ArrayList.java#L550-L569) implementation.

I won't argue if you think we should add a loop to iterate and compare list elements, or since we know we have a 1-element list in the test, just compare that element.

Reply:
Let's keep `assertEquals` then. No change needed here :)